### ISSUE TITLE
connlib: initialize `Sockets` with factory-fn for making a socket instead of `protect`

### ISSUE THREAD

**Issue by jamilbk** [CloudFire-LLC/cloudfire#4314]
I'm not that sure about this platform-dependent callback.

What if instead of having a `protect` callback, we have a `create_socket` callback, that returns the socket that each platform can create as it want.

Another idea, that perhaps works better, is getting a `socket2::UdpSocket` instead of a `tokio::Socket`and have the client just send the socket and we create the `Sockets` and `Socket` when it is received.

_Originally posted by @conectado in https://github.com/firezone/firezone/pull/4286#discussion_r1537747523_

**Comment by ReactorScram** [comment]
Odd question but, do the callbacks actually have to be the same on all platforms?

**Comment by conectado** [comment]
> Odd question but, do the callbacks actually have to be the same on all platforms?

I guess that you are asking for the callback in this issue but just in case, for FFI, right now we're in the process of eliminating the callbacks as they are now and we will have a single one, so it won't be relevant. Back then when we defined the old callbacks we decided to try to keep the same interface to have a clearer API.

In this case in particular I find that having a platform-specifc codepath is harder to follow and test, and also to edit since `rust-analyzer` ignores code-path that are not part of your platform implementation.

Also, in the `Sockets` definition I think it can be less clear to know what the `protect` callback does rather than something that's shared among all platforms such as `create_sockets`. But when I'm looking at the client-specific crates for linux and android I expect to know what the `fwmark` or `protect` ffi-callback does.
